keanu-project/src/test/java/io/improbable/keanu/vertices/intgr/probabilistic/MultinomialVertexTest.java

```diff
@@ -0,0 +1,302 @@
+        assertThat(samples, hasValue(0, 0, 100, 0));
+    }
+
+    @Test
+    public void whenKEqualsTwoItsBinomial() {
+        IntegerTensor n = IntegerTensor.scalar(10);
+        DoubleTensor p = DoubleTensor.create(0.2, 0.8).transpose();
+        DiscreteDistribution multinomial = Multinomial.withParameters(n, p);
+        DiscreteDistribution binomial = Binomial.withParameters(DoubleTensor.scalar(0.2), n);
+        for (int value : ImmutableList.of(1, 2, 9, 10)) {
+            DoubleTensor binomialLogProbs = binomial.logProb(IntegerTensor.scalar(value));
+            DoubleTensor multinomialLogProbs = multinomial.logProb(IntegerTensor.create(value, 10 - value).transpose()).transpose();
+            assertThat(multinomialLogProbs, allCloseTo(new Double(1e-6), binomialLogProbs));
+        }
+    }
+
+    enum Colours {
+        RED, GREEN, BLUE
+    }
+
+    @Test
+    public void whenKNEqualsOneItsCategorical() {
+        IntegerTensor n = IntegerTensor.scalar(1);
+        DoubleTensor p = DoubleTensor.create(0.2, .3, 0.5).transpose();
+        DiscreteDistribution multinomial = Multinomial.withParameters(n, p);
+
+        Map<Colours, DoubleVertex> selectableValues = ImmutableMap.of(
+            Colours.RED, ConstantVertex.of(p.getValue(0)),
+            Colours.GREEN, ConstantVertex.of(p.getValue(1)),
+            Colours.BLUE, ConstantVertex.of(p.getValue(2)));
+        Categorical categorical = Categorical.withParameters(selectableValues);
+
+        double pRed = categorical.logProb(Colours.RED);
+        assertThat(multinomial.logProb(IntegerTensor.create(1, 0, 0).transpose()).scalar(), closeTo(pRed, 1e-7));
+        double pGreen = categorical.logProb(Colours.GREEN);
+        assertThat(multinomial.logProb(IntegerTensor.create(0, 1, 0).transpose()).scalar(), closeTo(pGreen, 1e-7));
+        double pBlue = categorical.logProb(Colours.BLUE);
+        assertThat(multinomial.logProb(IntegerTensor.create(0, 0, 1).transpose()).scalar(), closeTo(pBlue, 1e-7));
+    }
+
+    @Test
+    public void samplingProducesRealisticMeanAndStandardDeviation() {
+        int N = 10000;
+        DoubleTensor p = DoubleTensor.create(0.1, 0.2, 0.3, 0.4).transpose();
+        IntegerTensor n = IntegerTensor.scalar(500);
+
+        MultinomialVertex vertex = new MultinomialVertex(
+            new int[]{1, N},
+            ConstantVertex.of(n),
+            ConstantVertex.of(p)
+        );
+
+        IntegerTensor samples = vertex.sample();
+        assertThat(samples, hasShape(4, N));
+
+        for (int i = 0; i < samples.getShape()[0]; i++) {
+            System.out.println(i);
+            IntegerTensor sample = samples.slice(0, i);
+            Double probability = p.slice(0, i).scalar();
+            double mean = sample.toDouble().average();
+            double std = sample.toDouble().standardDeviation();
+
+            double epsilonForMean = 0.5;
+            double epsilonForVariance = 5.;
```

Comment:
NB my error tolerances are pretty large - is this OK?

Reply:
Does it approach 0 as you take more samples? If so then it's fine.

Reply:
Yes.

I ran it for `N` = 1 million. It took 18 minutes but it gave me an accuracy of about 0.01 for mean and 0.4 for variance.